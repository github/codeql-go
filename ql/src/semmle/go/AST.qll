/**
 * Provides classes for working with AST nodes.
 */

import go

/**
 * An AST node.
 */
class AstNode extends @node, Locatable {
  /**
   * Gets the `i`th child node of this node.
   *
   * Note that the precise indices of child nodes are considered an implementation detail
   * and are subject to change without notice.
   */
  AstNode getChild(int i) {
    result = this.(ExprParent).getChildExpr(i) or
    result = this.(GoModExprParent).getChildGoModExpr(i) or
    result = this.(StmtParent).getChildStmt(i) or
    result = this.(DeclParent).getDecl(i) or
    result = this.(GenDecl).getSpec(i) or
    fields(result, this, i)
  }

  /**
   * Gets a child node of this node.
   */
  AstNode getAChild() { result = getChild(_) }

  /**
   * Gets the number of child nodes of this node.
   */
  int getNumChild() { result = count(getAChild()) }

  /** Gets the parent node of this AST node, if any. */
  AstNode getParent() { this = result.getAChild() }

  /** Gets the parent node of this AST node, but without crossing function boundaries. */
  private AstNode parentInSameFunction() {
    result = getParent() and
    not this instanceof FuncDef
  }

  /** Gets the innermost function definition to which this AST node belongs, if any. */
  FuncDef getEnclosingFunction() { result = getParent().parentInSameFunction*() }

  override string toString() { result = "AST node" }
}

/**
 * An AST node whose children include expressions.
 */
class ExprParent extends @exprparent, AstNode {
  /**
   * Gets the `i`th child expression of this node.
   *
   * Note that the precise indices of child expressions are considered an implementation detail
   * and are subject to change without notice.
   */
  Expr getChildExpr(int i) { exprs(result, _, this, i) }

  /**
   * Gets an expression that is a child node of this node in the AST.
   */
  Expr getAChildExpr() { result = getChildExpr(_) }

  /**
   * Gets the number of child expressions of this node.
   */
  int getNumChildExpr() { result = count(getAChildExpr()) }
}

/**
 * An AST node whose children include go.mod expressions.
 */
class GoModExprParent extends @modexprparent, AstNode {
  /**
   * Gets the `i`th child expression of this node.
   *
   * Note that the precise indices of child expressions are considered an implementation detail
   * and are subject to change without notice.
   */
  GoModExpr getChildGoModExpr(int i) { modexprs(result, _, this, i) }

  /**
   * Gets an expression that is a child node of this node in the AST.
   */
  GoModExpr getAChildGoModExpr() { result = getChildGoModExpr(_) }

  /**
   * Gets the number of child expressions of this node.
   */
  int getNumChildGoModExpr() { result = count(getAChildGoModExpr()) }
}

/**
 * An AST node whose children include statements.
 */
class StmtParent extends @stmtparent, AstNode {
  /**
   * Gets the `i`th child statement of this node.
   *
   * Note that the precise indices of child statements are considered an implementation detail
   * and are subject to change without notice.
   */
  Stmt getChildStmt(int i) { stmts(result, _, this, i) }

  /**
   * Gets a statement that is a child node of this node in the AST.
   */
  Stmt getAChildStmt() { result = getChildStmt(_) }

  /**
   * Gets the number of child statements of this node.
   */
  int getNumChildStmt() { result = count(getAChildStmt()) }
}

/**
 * An AST node whose children include declarations.
 */
class DeclParent extends @declparent, AstNode {
  /**
   * Gets the `i`th child declaration of this node.
   *
   * Note that the precise indices of declarations are considered an implementation detail
   * and are subject to change without notice.
   */
  Decl getDecl(int i) { decls(result, _, this, i) }

  /**
   * Gets a child declaration of this node in the AST.
   */
  Decl getADecl() { result = getDecl(_) }

  /**
   * Gets the number of child declarations of this node.
   */
  int getNumDecl() { result = count(getADecl()) }
}

/**
 * An AST node which may induce a scope.
 *
 * The following nodes may induce scopes:
 *
 *   - files
 *   - block statements, `if` statements, `switch` statements, `case` clauses, comm clauses, loops
 *   - function type expressions
 *
 * Note that functions themselves do not induce a scope, it is their type declaration that induces
 * the scope.
 */
class ScopeNode extends @scopenode, AstNode {
  /** Gets the scope induced by this node, if any. */
  LocalScope getScope() { scopenodes(this, result) }
}
