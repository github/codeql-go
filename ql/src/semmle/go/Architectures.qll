/** Provides classes for working with architectures. */

import go

/**
 * An architecture that is valid in a build constraint.
 *
 * Information obtained from
 * https://github.com/golang/go/blob/98cbf45cfc6a5a50cc6ac2367f9572cb198b57c7/src/go/types/gccgosizes.go
 * where the first field of the struct is 4 for 32-bit architectures
 * and 8 for 64-bit architectures.
 */
class Architecture extends string {
  int bitSize;

  Architecture() {
    this in [
        "386", "amd64p32", "arm", "armbe", "mips", "mipsle", "mips64p32", "mips64p32le", "ppc",
        "s390", "sparc"
      ] and
    bitSize = 32
    or
    this in [
        "amd64", "arm64", "arm64be", "ppc64", "ppc64le", "mips64", "mips64le", "s390x", "sparc64"
      ] and
    bitSize = 64
  }

  int getBitSize() { result = bitSize }
}
