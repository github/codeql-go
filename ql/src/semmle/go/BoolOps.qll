/**
 * Provides predicates and classes for working with boolean operations.
 */

import go

/** Provides predicates and classes for working with boolean operations. */
module BoolOps {
  /** BoolExpr is an expression that has boolean type. */
  class BoolExpr extends Expr {
    BoolExpr() { this.getType() instanceof BoolType }

    /**
     * Holds if the value of this node can influence `node`.
     *
     * Note that this does not hold if this is a parent of `node`.
     */
    predicate influences(BoolExpr node) {
      this = node
      or
      exists(BoolExpr mid | this.influences(mid) | node.(LogicalExpr).getAnOperand() = mid)
      or
      this.influences(node.(ParenExpr).getExpr())
    }

    /** Holds if, given that this node has value `value`, `node` must have value `ensured`. */
    predicate ensuresIf(Boolean value, BoolExpr node, boolean ensured) {
      this = node and value = ensured
      or
      exists(BoolExpr mid, boolean midVal | this.ensuresIf(value, mid, midVal) |
        // child -> parent, e.g. expr ensures (expr)
        node.(ParenExpr).getExpr() = mid and ensured = midVal
        or
        node.(NotExpr).getAnOperand() = mid and ensured = midVal.booleanNot()
        or
        node.(LandExpr).getAnOperand() = mid and midVal = false and ensured = false
        or
        node.(LorExpr).getAnOperand() = mid and midVal = true and ensured = true
        or
        // parent -> child, e.g. (expr) ensures expr
        node = mid.(ParenExpr).getExpr() and ensured = midVal
        or
        node = mid.(NotExpr).getAnOperand() and ensured = midVal.booleanNot()
        or
        node = mid.(LandExpr).getAnOperand() and midVal = true and ensured = true
        or
        node = mid.(LorExpr).getAnOperand() and midVal = false and ensured = false
      )
    }
  }
}
