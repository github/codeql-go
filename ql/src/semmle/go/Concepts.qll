/**
 * Provides abstract classes representing generic concepts such as file system
 * access or system command execution, for which individual framework libraries
 * provide concrete subclasses.
 */

import go
import semmle.go.dataflow.FunctionInputsAndOutputs

/**
 * A data-flow node that executes an operating system command,
 * for instance by spawning a new process.
 *
 * Extends this class to refine existing API models. If you want to model new APIs,
 * extend `SystemCommandExecution::Range` instead.
 */
class SystemCommandExecution extends DataFlow::Node {
  SystemCommandExecution::Range self;

  SystemCommandExecution() { this = self }

  /** Gets the argument that specifies the command to be executed. */
  DataFlow::Node getCommandName() { result = self.getCommandName() }
}

module SystemCommandExecution {
  /**
   * A data-flow node that executes an operating system command,
   * for instance by spawning a new process.
   *
   * Extend this class to model new APIs. If you want to refine existing API models,
   * extend `SystemCommandExecution` instead.
   */
  abstract class Range extends DataFlow::Node {
    /** Gets the argument that specifies the command to be executed. */
    abstract DataFlow::Node getCommandName();
  }
}

/**
 * An instantiation of a template; that is, a call which fills out a template with data.
 *
 * Extend this class to refine existing API models. If you want to model new APIs,
 * extend `TemplateInstantiation::Range` instead.
 */
class TemplateInstantiation extends DataFlow::Node {
  TemplateInstantiation::Range self;

  TemplateInstantiation() { this = self }

  /**
   * Gets the argument to this template instantiation that is the template being
   * instantiated.
   */
  DataFlow::Node getTemplateArgument() { result = self.getTemplateArgument() }

  /**
   * Gets an argument to this template instantiation that is data being inserted
   * into the template.
   */
  DataFlow::Node getADataArgument() { result = self.getADataArgument() }
}

module TemplateInstantiation {
  /**
   * An instantiation of a template; that is, a call which fills out a template with data.
   *
   * Extend this class to model new APIs. If you want to refine existing API models,
   * extend `TemplateInstantiation` instead.
   */
  abstract class Range extends DataFlow::Node {
    /**
     * Gets the argument to this template instantiation that is the template being
     * instantiated.
     */
    abstract DataFlow::Node getTemplateArgument();

    /**
     * Gets an argument to this template instantiation that is data being inserted
     * into the template.
     */
    abstract DataFlow::Node getADataArgument();
  }
}

/**
 * A data-flow node that performs a file system access, including reading and writing data,
 * creating and deleting files and folders, checking and updating permissions, and so on.
 *
 * Extend this class to refine existing API models. If you want to model new APIs,
 * extend `FileSystemAccess::Range` instead.
 */
class FileSystemAccess extends DataFlow::Node {
  FileSystemAccess::Range self;

  FileSystemAccess() { this = self }

  /** Gets an argument to this file system access that is interpreted as a path. */
  DataFlow::Node getAPathArgument() { result = self.getAPathArgument() }
}

module FileSystemAccess {
  /**
   * A data-flow node that performs a file system access, including reading and writing data,
   * creating and deleting files and folders, checking and updating permissions, and so on.
   *
   * Extend this class to model new APIs. If you want to refine existing API models,
   * extend `FileSystemAccess` instead.
   */
  abstract class Range extends DataFlow::Node {
    /** Gets an argument to this file system access that is interpreted as a path. */
    abstract DataFlow::Node getAPathArgument();
  }
}

/** A function that escapes meta-characters to prevent injection attacks. */
class EscapeFunction extends Function {
  EscapeFunction::Range self;

  EscapeFunction() { this = self }

  /**
   * The context that this function escapes for.
   *
   * Currently, this can be "js", "html", or "url".
   */
  string kind() { result = self.kind() }
}

module EscapeFunction {
  /**
   * A function that escapes meta-characters to prevent injection attacks.
   *
   * Extend this class to model new APIs. If you want to refine existing API models,
   * extend `EscapeFunction' instead.
   */
  abstract class Range extends Function {
    /**
     * The context that this function escapes for.
     *
     * Currently, this can be `js', `html', or `url'.
     */
    abstract string kind();
  }
}

/**
 * A function that escapes a string so it can be safely included in a
 * JavaScript string literal.
 */
class JsEscapeFunction extends EscapeFunction {
  JsEscapeFunction() { self.kind() = "js" }
}

/**
 * A function that escapes a string so it can be safely included in an
 * the body of an HTML element, for example, replacing `{}` in
 * `<p>{}</p>`.
 */
class HtmlEscapeFunction extends EscapeFunction {
  HtmlEscapeFunction() { self.kind() = "html" }
}

/**
 * A function that escapes a string so it can be safely included as part
 * of a URL.
 */
class UrlEscapeFunction extends EscapeFunction {
  UrlEscapeFunction() { self.kind() = "url" }
}

/**
 * A node whose value is interpreted as a part of a regular expression.
 *
 * Extend this class to refine existing API models. If you want to model new APIs,
 * extend `RegexpPattern::Range` instead.
 */
class RegexpPattern extends DataFlow::Node {
  RegexpPattern::Range self;

  RegexpPattern() { this = self }

  /**
   * Gets the node where this pattern is parsed as a part of a regular
   * expression.
   */
  DataFlow::Node getAParse() { result = self.getAParse() }

  /**
   * Gets this regexp pattern as a string.
   */
  string getPattern() { result = self.getPattern() }

  /**
   * Gets a use of this pattern, either as itself in an argument to a function or as a compiled
   * regexp object.
   */
  DataFlow::Node getAUse() { result = self.getAUse() }
}

module RegexpPattern {
  /**
   * A node whose value is interpreted as a part of a regular expression.
   *
   * Extend this class to model new APIs. If you want to refine existing API models,
   * extend `RegexpPattern' instead.
   */
  abstract class Range extends DataFlow::Node {
    /**
     * Gets a node where the pattern of this node is parsed as a part of
     * a regular expression.
     */
    abstract DataFlow::Node getAParse();

    /**
     * Gets this regexp pattern as a string.
     */
    abstract string getPattern();

    /**
     * Gets a use of this pattern, either as itself in an argument to a function or as a compiled
     * regexp object.
     */
    abstract DataFlow::Node getAUse();
  }
}

/**
 * A function that matches a regexp with a string or byte slice.
 *
 * Extend this class to refine existing API models. If you want to model new APIs,
 * extend `RegexpMatchFunction::Range` instead.
 */
class RegexpMatchFunction extends Function {
  RegexpMatchFunction::Range self;

  RegexpMatchFunction() { this = self }

  /**
   * Gets the function input that is the regexp being matched.
   */
  FunctionInput getRegexpArg() { result = self.getRegexpArg() }

  /**
   * Gets the regexp pattern that is used in the call to this function `call`.
   */
  RegexpPattern getRegexp(DataFlow::CallNode call) {
    result.getAUse() = this.getRegexpArg().getNode(call)
  }

  /**
   * Gets the function input that is the string being matched against.
   */
  FunctionInput getValue() { result = self.getValue() }

  /**
   * Gets the function output that is the Boolean result of the match function.
   */
  FunctionOutput getResult() { result = self.getResult() }
}

module RegexpMatchFunction {
  /**
   * A function that matches a regexp with a string or byte slice.
   *
   * Extend this class to model new APIs. If you want to refine existing API models,
   * extend `RegexpPattern' instead.
   */
  abstract class Range extends Function {
    /**
     * Gets the function input that is the regexp being matched.
     */
    abstract FunctionInput getRegexpArg();

    /**
     * Gets the function input that is the string being matched against.
     */
    abstract FunctionInput getValue();

    /**
     * Gets the Boolean result of the match function.
     */
    abstract FunctionOutput getResult();
  }
}

/**
 * A function that uses a regexp to replace parts of a string or byte slice.
 *
 * Extend this class to refine existing API models. If you want to model new APIs,
 * extend `RegexpReplaceFunction::Range` instead.
 */
class RegexpReplaceFunction extends Function {
  RegexpReplaceFunction::Range self;

  RegexpReplaceFunction() { this = self }

  /**
   * Gets the function input that is the regexp that matches text to replace.
   */
  FunctionInput getRegexpArg() { result = self.getRegexpArg() }

  /**
   * Gets the regexp pattern that is used to match patterns to replace in the call to this function
   * `call`.
   */
  RegexpPattern getRegexp(DataFlow::CallNode call) {
    result.getAUse() = call.(DataFlow::MethodCallNode).getReceiver()
  }

  /**
   * Gets the function input corresponding to the source value, that is, the value that is having
   * its contents replaced.
   */
  FunctionInput getSource() { result = self.getSource() }

  /**
   * Gets the function output corresponding to the result, that is, the value after replacement has
   * occurred.
   */
  FunctionOutput getResult() { result = self.getResult() }
}

module RegexpReplaceFunction {
  /**
   * A function that uses a regexp to replace parts of a string or byte slice.
   *
   * Extend this class to model new APIs. If you want to refine existing API models,
   * extend `RegexpReplaceFunction' instead.
   */
  abstract class Range extends Function {
    /**
     * Gets the function input that is the regexp that matches text to replace.
     */
    abstract FunctionInput getRegexpArg();

    /**
     * Gets the function input corresponding to the source value, that is, the value that is having
     * its contents replaced.
     */
    abstract FunctionInput getSource();

    /**
     * Gets the function output corresponding to the result, that is, the value after replacement
     * has occurred.
     */
    abstract FunctionOutput getResult();
  }
}

module HTTP {
  module ResponseWriter {
    /**
     * A variable that is an HTTP response writer.
     *
     * Extend this class to model new APIs. If you want to refine existing API models,
     * extend `HTTP::ResponseWriter` instead.
     */
    abstract class Range extends Variable {
      /** Gets a data-flow node that represents this response writer. */
      DataFlow::Node getANode() { result = this.getARead().getASuccessor*() }
    }
  }

  /**
   * A variable that is an HTTP response writer.
   *
   * Extend this class to refine existing API models. If you want to model new APIs,
   * extend `HTTP::ResponseWriter::Range` instead.
   */
  class ResponseWriter extends Variable {
    ResponseWriter::Range self;

    ResponseWriter() { this = self }

    /** Gets the body that is written in this HTTP response. */
    ResponseBody getBody() { result.getResponseWriter() = this }

    /** Gets a header write that is written in this HTTP response. */
    HeaderWrite getAHeaderWrite() { result.getResponseWriter() = this }

    /** Gets a redirect that is sent in this HTTP response. */
    Redirect getARedirect() { result.getResponseWriter() = this }

    /** Gets a data-flow node that represents this response writer. */
    DataFlow::Node getANode() { result = self.getANode() }
  }

  module HeaderWrite {
    /**
     * A data-flow node that represents a write to an HTTP header.
     *
     * Extend this class to model new APIs. If you want to refine existing API models,
     * extend `HTTP::HeaderWrite` instead.
     */
    abstract class Range extends DataFlow::ExprNode {
      /** Gets the (lower-case) name of a header set by this definition. */
      string getHeaderName() { result = this.getName().getStringValue().toLowerCase() }

      /** Holds if this header write defines the header `header`. */
      predicate definesHeader(string header, string value) {
        header = this.getName().getStringValue().toLowerCase() and
        value = this.getValue().getStringValue()
      }

      /** Gets the node representing the name of the header defined by this write. */
      abstract DataFlow::Node getName();

      /** Gets the node representing the value of the header defined by this write. */
      abstract DataFlow::Node getValue();

      /** Gets the response writer associated with this header write, if any. */
      abstract ResponseWriter getResponseWriter();
    }
  }

  /**
   * A data-flow node that represents a write to an HTTP header.
   *
   * Extend this class to refine existing API models. If you want to model new APIs,
   * extend `HTTP::HeaderWrite::Range` instead.
   */
  class HeaderWrite extends DataFlow::ExprNode {
    HeaderWrite::Range self;

    HeaderWrite() { this = self }

    /** Gets the (lower-case) name of a header set by this definition. */
    string getHeaderName() { result = self.getHeaderName() }

    /** Holds if this header write defines the header `header`. */
    predicate definesHeader(string header, string value) { self.definesHeader(header, value) }

    /** Gets the node representing the name of the header defined by this write. */
    DataFlow::Node getName() { result = self.getName() }

    /** Gets the node representing the value of the header defined by this write. */
    DataFlow::Node getValue() { result = self.getValue() }

    /** Gets the response writer associated with this header write, if any. */
    ResponseWriter getResponseWriter() { result = self.getResponseWriter() }
  }

  /** A data-flow node whose value is written to an HTTP header. */
  class Header extends DataFlow::Node {
    HeaderWrite hw;

    Header() {
      this = hw.getName()
      or
      this = hw.getValue()
    }

    /** Gets the response writer associated with this header write, if any. */
    ResponseWriter getResponseWriter() { result = hw.getResponseWriter() }
  }

  /** A data-flow node whose value is written to the value of an HTTP header. */
  class HeaderValue extends Header {
    HeaderValue() { this = hw.getValue() }
  }

  /** A data-flow node whose value is written to the name of an HTTP header. */
  class HeaderName extends Header {
    HeaderName() { this = hw.getName() }
  }

  module RequestBody {
    /**
     * An expression representing a reader whose content is written to an HTTP request body.
     *
     * Extend this class to model new APIs. If you want to refine existing API models,
     * extend `HTTP::RequestBody` instead.
     */
    abstract class Range extends DataFlow::Node { }
  }

  /**
   * An expression representing a reader whose content is written to an HTTP request body.
   *
   * Extend this class to refine existing API models. If you want to model new APIs,
   * extend `HTTP::RequestBody::Range` instead.
   */
  class RequestBody extends DataFlow::Node {
    RequestBody::Range self;

    RequestBody() { this = self }
  }

  module ResponseBody {
    /**
     * An expression which is written to an HTTP response body.
     *
     * Extend this class to model new APIs. If you want to refine existing API models,
     * extend `HTTP::ResponseBody` instead.
     */
    abstract class Range extends DataFlow::Node {
      /** Gets the response writer associated with this header write, if any. */
      abstract ResponseWriter getResponseWriter();
    }
  }

  /**
   * An expression which is written to an HTTP response body.
   *
   * Extend this class to refine existing API models. If you want to model new APIs,
   * extend `HTTP::ResponseBody::Range` instead.
   */
  class ResponseBody extends DataFlow::Node {
    ResponseBody::Range self;

    ResponseBody() { this = self }

    /** Gets the response writer associated with this header write, if any. */
    ResponseWriter getResponseWriter() { result = self.getResponseWriter() }
  }

  module Redirect {
    /**
     * An HTTP redirect.
     *
     * Extend this class to model new APIs. If you want to refine existing API models,
     * extend `HTTP::Redirect` instead.
     */
    abstract class Range extends DataFlow::Node {
      /** Gets the data-flow node representing the URL being redirected to. */
      abstract DataFlow::Node getUrl();

      /** Gets the response writer that this redirect is sent on. */
      abstract ResponseWriter getResponseWriter();
    }

    /**
     * An assignment of the HTTP Location header, which indicates the location for a
     * redirect.
     */
    private class LocationHeaderSet extends Range, HeaderWrite {
      LocationHeaderSet() { this.getHeaderName() = "location" }

      override DataFlow::Node getUrl() { result = this.getValue() }

      override ResponseWriter getResponseWriter() { result = HeaderWrite.super.getResponseWriter() }
    }
  }

  /**
   * An HTTP redirect.
   *
   * Extend this class to refine existing API models. If you want to model new APIs,
   * extend `HTTP::Redirect::Range` instead.
   */
  class Redirect extends DataFlow::Node {
    Redirect::Range self;

    Redirect() { this = self }

    /** Gets the data-flow node representing the URL being redirected to. */
    DataFlow::Node getUrl() { result = self.getUrl() }

    /** Gets the response writer that this redirect is sent on. */
    ResponseWriter getResponseWriter() { result = self.getResponseWriter() }
  }
}

/**
 * A call to a logging mechanism.
 *
 * Extend this class to refine existing API models. If you want to model new APIs,
 * extend `LoggerCall::Range` instead.
 */
class LoggerCall extends DataFlow::Node {
  LoggerCall::Range self;

  LoggerCall() { this = self }

  /** Gets a node that is a part of the logged message. */
  DataFlow::Node getAMessageComponent() { result = self.getAMessageComponent() }
}

module LoggerCall {
  /**
   * A call to a logging mechanism.
   *
   * Extend this class to model new APIs. If you want to refine existing API models,
   * extend `LoggerCall` instead.
   */
  abstract class Range extends DataFlow::Node {
    /** Gets a node that is a part of the logged message. */
    abstract DataFlow::Node getAMessageComponent();
  }
}
