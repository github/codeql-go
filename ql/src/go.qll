/**
 * Provides classes for working with Go programs.
 */

import Customizations
import semmle.go.AST
import semmle.go.Comments
import semmle.go.Concepts
import semmle.go.Decls
import semmle.go.Expr
import semmle.go.Files
import semmle.go.GoModExpr
import semmle.go.Locations
import semmle.go.Packages
import semmle.go.Scopes
import semmle.go.Stmt
import semmle.go.StringOps
import semmle.go.Types
import semmle.go.controlflow.BasicBlocks
import semmle.go.controlflow.ControlFlowGraph
import semmle.go.controlflow.IR
import semmle.go.dataflow.DataFlow
import semmle.go.dataflow.GlobalValueNumbering
import semmle.go.dataflow.TaintTracking
import semmle.go.dataflow.SSA
import semmle.go.frameworks.HTTP
import semmle.go.frameworks.SystemCommandExecutors
import semmle.go.frameworks.SQL
import semmle.go.frameworks.Stdlib
import semmle.go.security.FlowSources
import semmle.go.Util
