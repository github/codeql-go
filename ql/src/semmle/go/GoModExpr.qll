/**
 * Provides classes for working with go.mod expressions.
 */

import go

/**
 * A go.mod expression.
 */
class GoModExpr extends @modexpr, GoModExprParent {
  /**
   * Gets the kind of this expression, which is an integer value representing the expression's
   * node type.
   *
   * Note that the mapping from node types to integer kinds is considered an implementation detail
   * and subject to change without notice.
   */
  int getKind() { modexprs(this, result, _, _) }

  /**
   * Get the comment group associated with this expression.
   */
  DocComment getComments() { result.getDocumentedElement() = this }

  override string toString() { result = "go.mod expression" }
}

/**
 *  A top-level block of comments separate from any rule.
 */
class GoModCommentBlock extends @modcommentblock, GoModExpr { }

/**
 * A single line of tokens.
 */
class GoModLine extends @modline, GoModExpr {
  /**
   * Gets the `i`th token on this line.
   */
  string getToken(int i) { modtokens(result, this, i) }

  override string toString() { result = "go.mod line" }
}

/**
 * A line that contains the module information
 */
class GoModModuleLine extends GoModLine {
  GoModModuleLine() {
    this.getParent().(GoModLineBlock).getToken(0) = "module"
    or
    not this.getParent() instanceof GoModLineBlock and
    this.getToken(0) = "module"
  }

  string getPath() {
    if this.getParent() instanceof GoModLineBlock
    then result = this.getToken(1)
    else result = this.getToken(0)
  }

  override string toString() { result = "go.mod module line" }
}

/**
 * A factored block of lines, for example `require ( "path" )`.
 */
class GoModLineBlock extends @modlineblock, GoModExpr {
  /**
   * Gets the `i`th token of this line block.
   */
  string getToken(int i) { modtokens(result, this, i) }

  override string toString() { result = "go.mod line block" }
}

/**
 * A line that declares the Go version to be used, for example `go 1.14`.
 */
class GoModGoLine extends GoModLine {
  GoModGoLine() { this.getToken(0) = "go" }

  /** Gets the Go version declared. */
  string getVer() { result = this.getToken(1) }

  override string toString() { result = "go.mod go line" }
}

private string getOffsetToken(GoModLine line, int i) {
  if line.getParent() instanceof GoModLineBlock
  then result = line.getToken(i - 1)
  else result = line.getToken(i)
}

/**
 * A line that declares a requirement, for example `require "path"`.
 */
class GoModRequireLine extends GoModLine {
  GoModRequireLine() {
    this.getParent().(GoModLineBlock).getToken(0) = "require"
    or
    not this.getParent() instanceof GoModLineBlock and
    this.getToken(0) = "require"
  }

  /** Gets the path of the dependency. */
  string getPath() { result = getOffsetToken(this, 1) }

  /** Gets the version of the dependency. */
  string getVer() { result = getOffsetToken(this, 2) }

  override string toString() { result = "go.mod require line" }
}

/**
 * A line that declares a dependency version to exclude, for example `exclude "ver"`.
 */
class GoModExcludeLine extends GoModLine {
  GoModExcludeLine() {
    this.getParent().(GoModLineBlock).getToken(0) = "exclude"
    or
    not this.getParent() instanceof GoModLineBlock and
    this.getToken(0) = "exclude"
  }

  /** Gets the path of the dependency to exclude a version of. */
  string getPath() { result = getOffsetToken(this, 1) }

  /** Gets the excluded version. */
  string getVer() { result = getOffsetToken(this, 2) }

  override string toString() { result = "go.mod exclude line" }
}

/**
 * A line that specifies a dependency to use instead of another one, for example
 * `replace "a" => "b" ver`.
 */
class GoModReplaceLine extends GoModLine {
  GoModReplaceLine() {
    this.getParent().(GoModLineBlock).getToken(0) = "replace"
    or
    not this.getParent() instanceof GoModLineBlock and
    this.getToken(0) = "replace"
  }

  /** Gets the path of the dependency to be replaced. */
  string getOriginalPath() { result = getOffsetToken(this, 1) }

  /** Gets the path of the replacement dependency. */
  string getReplacementPath() { result = getOffsetToken(this, 3) }

  /** Gets the version of the replacement dependency. */
  string getReplacementVer() { result = getOffsetToken(this, 4) }

  override string toString() { result = "go.mod replace line" }
}

/** A left parenthesis for a line block. */
class GoModLParen extends @modlparen, GoModExpr {
  override string toString() { result = "go.mod (" }
}

/** A right parenthesis for a line block. */
class GoModRParen extends @modrparen, GoModExpr {
  override string toString() { result = "go.mod )" }
}
