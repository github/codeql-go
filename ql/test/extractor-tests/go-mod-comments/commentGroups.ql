import go

from CommentGroup cg
select cg
