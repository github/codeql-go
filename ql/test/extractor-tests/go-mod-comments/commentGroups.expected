| go.mod:1:1:2:32 | comment group |
| go.mod:3:1:3:11 | comment group |
| go.mod:5:1:5:13 | comment group |
